workers/unity/Assets/Playground/Scripts/Worker/WorkerConnectorBase.cs

```diff
@@ -0,0 +1,225 @@
+﻿using System;
+using System.Collections;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Improbable.Gdk.Core;
+using Improbable.Worker.Core;
+using Unity.Entities;
+using UnityEngine;
+
+namespace Playground
+{
+    public class WorkerConnectorBase : MonoBehaviour, IDisposable
+    {
+        private delegate Task<Worker> ConnectionDelegate();
+
+        public GameObject LevelPrefab;
+        public int MaxConnectionAttempts = 3;
+        public bool UseExternalIp = false;
+
+        public Worker Worker;
+
+        protected WorkerConnectorBase[] RequiredWorkerConnection;
+
+        private GameObject levelInstance;
+
+        private readonly TaskCompletionSource<bool> connectionAttemptFinishedTask = new TaskCompletionSource<bool>();
+
+        // Important run in this step as otherwise it can interfere with the the domain unloading logic
+        private void OnApplicationQuit()
+        {
+            Dispose();
+        }
+
+        private void OnDestroy()
+        {
+            Dispose();
+        }
+
+        public async Task Connect(string workerType, ILogDispatcher logger)
+        {
+            // Check that other workers have finished trying to connect before this one starts
+            // This prevents races on the workers starting and races on when we start ticking systems
+            if (RequiredWorkerConnection != null)
+            {
+                Task[] requiredWorkers = RequiredWorkerConnection
+                    .Select(connection => connection.connectionAttemptFinishedTask.Task).ToArray();
+                try
+                {
+                    await Task.WhenAll(requiredWorkers);
+                }
+                catch (Exception e)
+                {
+                    logger.HandleLog(LogType.Error,
+                        new LogEvent("Cancelling connection attempt")
+                            .WithField("WorkerType", workerType)
+                            .WithField("Reason", "Required worker failed to connect"));
+                    connectionAttemptFinishedTask.SetException(e);
+                    Dispose();
+                    return;
+                }
+            }
+
+            try
+            {
+                var origin = transform.position;
+                ConnectionDelegate connectionDelegate;
+                if (ShouldUseLocator())
+                {
+                    var config = GetLocatorConfig(workerType);
+                    connectionDelegate = async () =>
```

Comment:
These are a bit of an eyeful, but could be dealt with at a later date.

Reply:
We can make them inline functions or make them point to functions.